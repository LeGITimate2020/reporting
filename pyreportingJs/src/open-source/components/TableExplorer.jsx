import React from 'react';
import PropTypes from 'prop-types';
import { Tab, Tabs } from 'react-bootstrap';
import _ from 'lodash';
import CrossFilterPanel from './CrossFilterPanel';
import PCAPanel from './PCAPanel';
import { convertRecordArraysToRecordObjects, runPCA } from '../utils/dataManipulationHelpers';

import Table from './Table';

/**
 * Pure component to generate tab strip based on props
 */
const TableExplorerTabStrip = (props) => {
  const {
    data, pcaData, reactTableColumns, title, fileId,
  } = props;
  if (data.length === 0) {
    return null;
  }

  return (
    <Tabs animation={false} className="csv-panel-tab" id="csv-panel-tabs">
      <Tab eventKey={1} title="Table">
        <Table
          key={fileId}
          title={title}
          columns={reactTableColumns}
          data={data}
          colorByColumn
          reactTableProps={{ defaultPageSize: 25 }}
        />
      </Tab>
      <Tab eventKey={2} title="Linked Scatter">
        <div>
          {data.length > 0 && <CrossFilterPanel key={fileId} data={data} title={title} />}
        </div>
      </Tab>
      <Tab eventKey={3} title="Principal Component Analysis">
        <div>
          <PCAPanel {...pcaData} />
        </div>
      </Tab>
    </Tabs>
  );
};

/**
 * Tabular data explorer:
 * 1) Sortable table
 * 2) Linked Scatter
 * 3) Optional PCA analysis
 *
 * Data format:
 * [
 *   ['header1', 'header2', 'header3'],
 *   [11, 21, 31],
 *   [12, 22, 32],
 *   [13, 23, 33],
 * ];
 */
export default class TableExplorer extends React.Component {
  static propTypes = {
    data: PropTypes.array,
    shouldRunPCA: PropTypes.bool,
    title: PropTypes.string,
  };

  static defaultProps = {
    data: [],
    shouldRunPCA: false,
    title: '',
  };

  constructor(props) {
    super(props);

    this.state = {
      shouldRunPCA: !!this.props.shouldRunPCA,
      fileId: 0,
    };
  }

  _derivePropsFromData = (data, shouldRunPCA) => {
    let recordArrays;
    let pcaData = {};

    if (shouldRunPCA) {
      [recordArrays, pcaData] = runPCA(data);
    } else {
      recordArrays = data;
    }

    const columns = recordArrays[0];

    // convert to a format usable by <ReactTable />
    const reactTableColumns = columns.map(column => ({ Header: column, accessor: column }));
    return {
      columns,
      reactTableColumns,
      data: convertRecordArraysToRecordObjects(recordArrays),
      pcaData,
    };
  };

  onPCAClick = () => {
    this.setState(state => ({
      shouldRunPCA: true,
      fileId: state.fileId + 1,
    }));
  };

  render() {
    const { data, title } = this.props;
    if (data.length === 0) {
      return null;
    }
    const { fileId, shouldRunPCA } = this.state;

    // In order to minimize component state, it is fine to do data manipulation here in the render method. The important
    // thing to keep in mind is to make sure there are no unnecessary renders.
    const derivedProps = this._derivePropsFromData(data, shouldRunPCA);

    return (
      <div className="row">
        <div className="col-lg-9">
          {
            _.isEmpty(derivedProps.pcaData)
            && (
            <button
              className="btn btn-sm btn-info btn-run-pca"
              onClick={this.onPCAClick}
            >
              Run PCA
            </button>
            )
          }
        </div>
        <div className="col-lg-12">
          <TableExplorerTabStrip title={title} fileId={fileId} {...derivedProps} />
        </div>
      </div>
    );
  }
}
