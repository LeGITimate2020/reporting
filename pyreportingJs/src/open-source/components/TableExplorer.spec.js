import React from 'react';
import { shallow } from 'enzyme';
import TableExplorer from './TableExplorer';

describe('TableExplorer.jsx', () => {
  const csvData = [
    ['header1', 'header2', 'header3'],
    [11, 21, 31],
    [12, 22, 32],
    [13, 23, 33],
  ];

  const props = { data: csvData };

  it('renders without crashing', () => {
    shallow(<TableExplorer />);
  });

  it('Can run PCA on data from props', () => {
    const wrapper = shallow(<TableExplorer {...props} />);
    const btnWrapper = wrapper.find('.btn-run-pca');

    // Button exists because data was successfully imported (without drag-and-drop)
    expect(btnWrapper.length).toBe(1);
    btnWrapper.simulate('click');

    // Button disappears after running PCA
    expect(wrapper.find('.btn-run-pca').length).toBe(0);
  });

  it('Matches snapshot', () => {
    const wrapper = shallow(<TableExplorer {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
