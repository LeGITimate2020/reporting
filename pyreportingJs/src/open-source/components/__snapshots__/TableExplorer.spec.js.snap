// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableExplorer.jsx Matches snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TableExplorer
    data={
      Array [
        Array [
          "header1",
          "header2",
          "header3",
        ],
        Array [
          11,
          21,
          31,
        ],
        Array [
          12,
          22,
          32,
        ],
        Array [
          13,
          23,
          33,
        ],
      ]
    }
    shouldRunPCA={false}
    title=""
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="col-lg-9"
        >
          <button
            className="btn btn-sm btn-info btn-run-pca"
            onClick={[Function]}
          >
            Run PCA
          </button>
        </div>,
        <div
          className="col-lg-12"
        >
          <TableExplorerTabStrip
            columns={
              Array [
                "header1",
                "header2",
                "header3",
              ]
            }
            data={
              Array [
                Object {
                  "header1": 11,
                  "header2": 21,
                  "header3": 31,
                },
                Object {
                  "header1": 12,
                  "header2": 22,
                  "header3": 32,
                },
                Object {
                  "header1": 13,
                  "header2": 23,
                  "header3": 33,
                },
              ]
            }
            fileId={0}
            pcaData={Object {}}
            reactTableColumns={
              Array [
                Object {
                  "Header": "header1",
                  "accessor": "header1",
                },
                Object {
                  "Header": "header2",
                  "accessor": "header2",
                },
                Object {
                  "Header": "header3",
                  "accessor": "header3",
                },
              ]
            }
            title=""
          />
        </div>,
      ],
      "className": "row",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <button
            className="btn btn-sm btn-info btn-run-pca"
            onClick={[Function]}
          >
            Run PCA
          </button>,
          "className": "col-lg-9",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Run PCA",
            "className": "btn btn-sm btn-info btn-run-pca",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Run PCA",
          "type": "button",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <TableExplorerTabStrip
            columns={
              Array [
                "header1",
                "header2",
                "header3",
              ]
            }
            data={
              Array [
                Object {
                  "header1": 11,
                  "header2": 21,
                  "header3": 31,
                },
                Object {
                  "header1": 12,
                  "header2": 22,
                  "header3": 32,
                },
                Object {
                  "header1": 13,
                  "header2": 23,
                  "header3": 33,
                },
              ]
            }
            fileId={0}
            pcaData={Object {}}
            reactTableColumns={
              Array [
                Object {
                  "Header": "header1",
                  "accessor": "header1",
                },
                Object {
                  "Header": "header2",
                  "accessor": "header2",
                },
                Object {
                  "Header": "header3",
                  "accessor": "header3",
                },
              ]
            }
            title=""
          />,
          "className": "col-lg-12",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "columns": Array [
              "header1",
              "header2",
              "header3",
            ],
            "data": Array [
              Object {
                "header1": 11,
                "header2": 21,
                "header3": 31,
              },
              Object {
                "header1": 12,
                "header2": 22,
                "header3": 32,
              },
              Object {
                "header1": 13,
                "header2": 23,
                "header3": 33,
              },
            ],
            "fileId": 0,
            "pcaData": Object {},
            "reactTableColumns": Array [
              Object {
                "Header": "header1",
                "accessor": "header1",
              },
              Object {
                "Header": "header2",
                "accessor": "header2",
              },
              Object {
                "Header": "header3",
                "accessor": "header3",
              },
            ],
            "title": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="col-lg-9"
          >
            <button
              className="btn btn-sm btn-info btn-run-pca"
              onClick={[Function]}
            >
              Run PCA
            </button>
          </div>,
          <div
            className="col-lg-12"
          >
            <TableExplorerTabStrip
              columns={
                Array [
                  "header1",
                  "header2",
                  "header3",
                ]
              }
              data={
                Array [
                  Object {
                    "header1": 11,
                    "header2": 21,
                    "header3": 31,
                  },
                  Object {
                    "header1": 12,
                    "header2": 22,
                    "header3": 32,
                  },
                  Object {
                    "header1": 13,
                    "header2": 23,
                    "header3": 33,
                  },
                ]
              }
              fileId={0}
              pcaData={Object {}}
              reactTableColumns={
                Array [
                  Object {
                    "Header": "header1",
                    "accessor": "header1",
                  },
                  Object {
                    "Header": "header2",
                    "accessor": "header2",
                  },
                  Object {
                    "Header": "header3",
                    "accessor": "header3",
                  },
                ]
              }
              title=""
            />
          </div>,
        ],
        "className": "row",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <button
              className="btn btn-sm btn-info btn-run-pca"
              onClick={[Function]}
            >
              Run PCA
            </button>,
            "className": "col-lg-9",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Run PCA",
              "className": "btn btn-sm btn-info btn-run-pca",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Run PCA",
            "type": "button",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <TableExplorerTabStrip
              columns={
                Array [
                  "header1",
                  "header2",
                  "header3",
                ]
              }
              data={
                Array [
                  Object {
                    "header1": 11,
                    "header2": 21,
                    "header3": 31,
                  },
                  Object {
                    "header1": 12,
                    "header2": 22,
                    "header3": 32,
                  },
                  Object {
                    "header1": 13,
                    "header2": 23,
                    "header3": 33,
                  },
                ]
              }
              fileId={0}
              pcaData={Object {}}
              reactTableColumns={
                Array [
                  Object {
                    "Header": "header1",
                    "accessor": "header1",
                  },
                  Object {
                    "Header": "header2",
                    "accessor": "header2",
                  },
                  Object {
                    "Header": "header3",
                    "accessor": "header3",
                  },
                ]
              }
              title=""
            />,
            "className": "col-lg-12",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "columns": Array [
                "header1",
                "header2",
                "header3",
              ],
              "data": Array [
                Object {
                  "header1": 11,
                  "header2": 21,
                  "header3": 31,
                },
                Object {
                  "header1": 12,
                  "header2": 22,
                  "header3": 32,
                },
                Object {
                  "header1": 13,
                  "header2": 23,
                  "header3": 33,
                },
              ],
              "fileId": 0,
              "pcaData": Object {},
              "reactTableColumns": Array [
                Object {
                  "Header": "header1",
                  "accessor": "header1",
                },
                Object {
                  "Header": "header2",
                  "accessor": "header2",
                },
                Object {
                  "Header": "header3",
                  "accessor": "header3",
                },
              ],
              "title": "",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
